src/audio.c

```diff
@@ -0,0 +1,212 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with ReplaySorcery.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include "audio.h"
+#include "util/log.h"
+#include "util/memory.h"
+#include "pthread.h"
+#define SAMPLES_PER_CALLBACK 4096 //has to be power of 2
+static pthread_spinlock_t sampleGetLock;
+
+static void audioCallback(void *userdata, uint8_t *stream, int len) {
+   pthread_spin_lock(&sampleGetLock);
+   RSAudio* audio = (RSAudio*)userdata;
+   int end = audio->index + len;
+   if (end >= audio->size) {
+      audio->index = 0;
+   }
+   memcpy(audio->data + audio->index, stream, (size_t)len);
+   audio->index += len;
+   pthread_spin_unlock(&sampleGetLock);
+}
+
+static const char *getDeviceName(const RSConfig *config) {
+   const char* devname = config->audioDeviceName;
+   if (!devname || !strcmp(config->audioDeviceName, "auto")) {
+      return 0;
+   }
+   return devname;
+}
+
+static void printDevices(void) {
+   rsLog("Available devices:");
+   int max = SDL_GetNumAudioDevices(1);
+   for (int i = 0; i < max; ++i) {
+      rsLog(SDL_GetAudioDeviceName(i, 1));
+   }
+}
+
+static void prepareEncoder(RSAudio *audio, const RSConfig* config) {
+   RSAudioEncoder* audioenc = &audio->audioenc;
+   if (AACENC_OK != aacEncOpen(&(audioenc->aac_enc), 0, 0)) {
+      rsError("aacEncOpen failed");
+   }
+   aacEncoder_SetParam(audioenc->aac_enc, AACENC_TRANSMUX, 0);
+   aacEncoder_SetParam(audioenc->aac_enc, AACENC_AFTERBURNER, 1);
+   aacEncoder_SetParam(audioenc->aac_enc, AACENC_BITRATE, (UINT)config->audioBitrate);
+   aacEncoder_SetParam(audioenc->aac_enc, AACENC_SAMPLERATE, (UINT)config->audioSamplerate);
+   aacEncoder_SetParam(audioenc->aac_enc, AACENC_CHANNELMODE, (UINT)config->audioChannels);
+   if (AACENC_OK != aacEncEncode(audioenc->aac_enc, NULL, NULL, NULL, NULL)) {
+      rsError("aacEncEncode failed. Probably bad bitrate. Tried %d", config->audioBitrate);
+   }
+   aacEncInfo(audioenc->aac_enc, &(audio->audioenc.aac_info));
+
+   audioenc->data = 0;
+   audioenc->frame = 0;
+   audioenc->index = 0;
+   audioenc->size = audio->size;
+   audioenc->samplesPerFrame = (int)audioenc->aac_info.frameLength * config->audioChannels;
+   audioenc->frameSize = audioenc->samplesPerFrame * (int)sizeof(uint16_t);
+}
+
+static void prepareFrame(RSAudioEncoder *audioenc) {
+   int firstCopySize = (int)audioenc->frameSize;
+   int endData = (int)audioenc->index + audioenc->frameSize;
+   int diff = audioenc->size - endData;
+   if (diff >= 0) {
+      memcpy(audioenc->frame, audioenc->data + audioenc->index, (size_t)firstCopySize);
+      audioenc->index += firstCopySize;
+      return;
+   }
+   firstCopySize += diff;
+   memcpy(audioenc->frame, audioenc->data + audioenc->index, (size_t)firstCopySize);
+   memcpy(audioenc->frame + firstCopySize, audioenc->data, (size_t)-diff);
+   audioenc->index = -diff;
```

Comment:
Would be nice as an external util file like I said in the last PR

Reply:
workin on it

Reply:
done